### PR TITLE
Use PartitionsSpec for all task types

### PR DESCRIPTION
**Pull request by jihoonson**
Part of https://github.com/apache/incubator-druid/issues/8061.

### Description

`PartitionsSpec` is a class to describe the secondary partitioning method for data ingestion, but is being used by only Hadoop tasks. For more consistent behavior and configuration, all task types should use the same `PartitionsSpec`.

`PartitionsSpec` is the top interface and has one direct implementation, `DynamicPartitionsSpec`. `DynamicPartitionsSpec` is the new partitionsSpec and used by indexTask and kafka/kinesis IndexTasks.

`DimensionBasedPartitionsSpec` is the child interface of `PartitionsSpec` and represents the partitionsSpec based on dimension values. It has two implementations of `HashedPartitionsSpec` and `SingleDimensionPartitionsSpec`. These partitionsSpecs are used if and only if perfect rollup is configured.

This PR is _backward-Incompatible_ for indexTask because the JSON form of tuningConfig doesn't have `maxRowsPerSegment`, `maxTotalRows`, `numShards`, and `partitionDimensions` anymore. However, it still could read the old JSON format. It should be compatible for other task types.

<hr>

This PR has:
- [x] been self-reviewed.
- [x] added documentation for new or modified features or behaviors.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.

### CODE REVIEW

File: indexing-hadoop/src/main/java/org/apache/druid/indexer/partitions/PartitionsSpec.java [hunk 1]
```diff
@@ -19,40 +19,55 @@
...
+   * Implementations can have different default values which it could be even null.
+   * Callers should use the right value depending on the context if this returns null.
+   */
+  @Nullable
+  Integer getMaxRowsPerSegment();
 
-  @JsonIgnore
-  boolean isDeterminingPartitions();
+  /**
+   * Returns true if this partitionsSpec needs to determine the number of partitions to start data ingetsion.
```
Comment (clintropolis) [inline indexing-hadoop/src/main/java/org/apache/druid/indexer/partitions/PartitionsSpec.java]: typo: ingetsion

File: indexing-service/src/main/java/org/apache/druid/indexing/common/task/IndexTask.java [hunk 2]
```diff
@@ -1439,38 +1442,12 @@ public IndexTuningConfig withBasePersistDirectory(File dir)
...
-          basePersistDirectory,
-          segmentWriteOutMediumFactory,
-          logParseExceptions,
-          maxParseExceptions,
-          maxSavedParseExceptions
-      );
-    }
-
-    public IndexTuningConfig withMaxRowsPerSegment(int maxRowsPerSegment)
+    public IndexTuningConfig withPartitoinsSpec(PartitionsSpec partitionsSpec)
```
Comment (clintropolis) [inline indexing-service/src/main/java/org/apache/druid/indexing/common/task/IndexTask.java]: typo: `withPartitoinsSpec`

File: indexing-service/src/main/java/org/apache/druid/indexing/common/task/AppenderatorDriverRealtimeIndexTask.java [hunk 3]
```diff
@@ -316,7 +316,10 @@ public TaskStatus run(final TaskToolbox toolbox)
             AppenderatorDriverAddResult addResult = driver.add(inputRow, sequenceName, committerSupplier);
 
             if (addResult.isOk()) {
-              if (addResult.isPushRequired(tuningConfig)) {
+              final boolean isPushRequired =
```
Comment (clintropolis) [inline indexing-service/src/main/java/org/apache/druid/indexing/common/task/AppenderatorDriverRealtimeIndexTask.java]: Any reason not to push this down into the `AppenderatorDriverAddResult` like it was previously? Is the other `AppenderatorDriverAddResult.isPushRequired` method still legitimately used?

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java [hunk 4]
```diff
@@ -595,7 +595,10 @@ public void run()
                     if (addResult.isOk()) {
                       // If the number of rows in the segment exceeds the threshold after adding a row,
                       // move the segment out from the active segments of BaseAppenderatorDriver to make a new segment.
-                      if (addResult.isPushRequired(tuningConfig) && !sequenceToUse.isCheckpointed()) {
+                      final boolean isPushRequired =
```
Comment (clintropolis) [inline indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java]: same thing re `isPushRequired`

File: indexing-hadoop/src/main/java/org/apache/druid/indexer/partitions/PartitionsSpec.java (hunk 1 above)
Comment (jihoonson) [reply indexing-hadoop/src/main/java/org/apache/druid/indexer/partitions/PartitionsSpec.java]: Thanks, fixed.

File: indexing-service/src/main/java/org/apache/druid/indexing/common/task/IndexTask.java (hunk 2 above)
Comment (jihoonson) [reply indexing-service/src/main/java/org/apache/druid/indexing/common/task/IndexTask.java]: Thanks, fixed.

File: indexing-service/src/main/java/org/apache/druid/indexing/common/task/AppenderatorDriverRealtimeIndexTask.java (hunk 3 above)
Comment (jihoonson) [reply indexing-service/src/main/java/org/apache/druid/indexing/common/task/AppenderatorDriverRealtimeIndexTask.java]: Hmm, yeah it looks better to use the existing one. Reverted to use it.

File: indexing-service/src/main/java/org/apache/druid/indexing/common/task/IndexTask.java [hunk 5]
```diff
@@ -1243,21 +1217,60 @@ public static IndexTuningConfig createDefault()
...
+          if (maxRowsPerSegment != null || maxTotalRows != null) {
+            return new DynamicPartitionsSpec(maxRowsPerSegment, maxTotalRows);
+          } else {
+            return null;
+          }
+        }
+      } else {
+        if (forceGuaranteedRollup) {
+          if (!(partitionsSpec instanceof HashedPartitionsSpec)) {
+            throw new ISE("HashedPartitonsSpec must be used for perfect rollup");
```
Comment (dclim) [inline indexing-service/src/main/java/org/apache/druid/indexing/common/task/IndexTask.java]: Typo: `HashedPartitionsSpec`

File: indexing-hadoop/src/main/java/org/apache/druid/indexer/partitions/HashedPartitionsSpec.java [hunk 6]
```diff
@@ -20,45 +20,103 @@
...
+      @JsonProperty("maxRowsPerSegment") @Nullable Integer maxRowsPerSegment,
       @JsonProperty("numShards") @Nullable Integer numShards,
       @JsonProperty("partitionDimensions") @Nullable List<String> partitionDimensions
   )
   {
-    super(targetPartitionSize, maxPartitionSize, assumeGrouped, numShards);
-    this.partitionDimensions = partitionDimensions == null ? DEFAULT_PARTITION_DIMENSIONS : partitionDimensions;
+    Preconditions.checkArgument(
+        PartitionsSpec.isEffectivelyNull(maxRowsPerSegment) || PartitionsSpec.isEffectivelyNull(numShards),
+        "Can't use maxRowsPerSegment and numShards together"
```
[the remark this answers is not in the document]
Comment (jihoonson) [reply indexing-hadoop/src/main/java/org/apache/druid/indexer/partitions/HashedPartitionsSpec.java]: 👍

File: indexing-hadoop/src/main/java/org/apache/druid/indexer/partitions/HashedPartitionsSpec.java [hunk 7]
```diff
@@ -20,45 +20,103 @@
...
+    );
+    // Needs to determin partitions if the _given_ numShards is null
```
[the remark this answers is not in the document]
Comment (jihoonson) [reply indexing-hadoop/src/main/java/org/apache/druid/indexer/partitions/HashedPartitionsSpec.java]: Thanks, fixed.

File: indexing-hadoop/src/main/java/org/apache/druid/indexer/partitions/SingleDimensionPartitionsSpec.java [hunk 8]
```diff
@@ -22,29 +22,55 @@
...
-      @JsonProperty("maxPartitionSize") @Nullable Long maxPartitionSize,
-      @JsonProperty("assumeGrouped") @Nullable Boolean assumeGrouped
+      @JsonProperty("assumeGrouped") boolean assumeGrouped // false by default
   )
   {
-    super(targetPartitionSize, maxPartitionSize, assumeGrouped, null);
+    Preconditions.checkArgument(maxRowsPerSegment > 0, "maxRowsPerSegment must be specified");
+    this.maxRowsPerSegment = maxRowsPerSegment;
+    this.maxPartitionSize = PartitionsSpec.isEffectivelyNull(maxPartitionSize)
+                            ? Math.multiplyExact(maxRowsPerSegment, (int) (maxRowsPerSegment * 0.5))
```
[the remark this answers is not in the document]
Comment (jihoonson) [reply indexing-hadoop/src/main/java/org/apache/druid/indexer/partitions/SingleDimensionPartitionsSpec.java]: Nice finding! This should be `Math.addExact(maxRowsPerSegment, (int) (maxRowsPerSegment * 0.5))`. Thanks.

File: indexing-service/src/main/java/org/apache/druid/indexing/common/task/IndexTask.java (hunk 5 above)
Comment (jihoonson) [reply indexing-service/src/main/java/org/apache/druid/indexing/common/task/IndexTask.java]: Thanks, fixed.

**Comment by dclim** [comment]
👍 after conflict resolved

**Comment by jihoonson** [comment]
@dclim @clintropolis thank you for the review!

**Pull request merged**